### ISSUE TITLE
Higher memory usage because of #1732 and #1901

### ISSUE THREAD

**Issue by macekmar** [netket/netket#1956]
I noticed that change 3.13.0 → 3.14.0 uses more RAM.

The following example:
```python
import netket as nk
L = 20
g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)
hi = nk.hilbert.Spin(s=1 / 2, N=g.n_nodes)
ha = nk.operator.Ising(hilbert=hi, graph=g, h=1.0)
ha.to_sparse()
```
now crashes on a node with 64 GB of RAM.

The issue arises when we call `hilbert.states_to_local_indices` and `hilbert.local_indices_to_states`:

In `HomogeneusHilbert.states_to_local_indices` ([L202](https://github.com/netket/netket/blob/781f2d2bd3b3eda6ebb6eb13469fa78aed9d20f5/netket/hilbert/homogeneous.py#L202))we do
```
self._local_states.states_to_numbers(x, dtype=np.int32)
```
which in case of `SpinHilbert` transforms an `int8` into `int32` array (x4 memory)
Next, in `StaticRange.states_to_numbers` ([L152](https://github.com/netket/netket/blob/781f2d2bd3b3eda6ebb6eb13469fa78aed9d20f5/netket/utils/static_range.py#L152)) we do
```
idx = (x - self.start) / self.step
```
which transforms array `x` of type `int32` to `float64` (x8 memory than the original array). Note, that arrays `x` are different arrays in different scopes.
I think the reason why v3.13 is still working when #1732 is part of v3.12 is that `_ising/numba.py` was rewritten from 3.13.0 to 3.14.0.

When calculating `ha.to_sparse()`, the array `idx` took 40 % of all the memory: $8·n_\mathrm{samples}·L·n_\mathrm{conn} = 8 · 2^L · L · (L+1)$ Bytes of memory (times 16 cores¹). This means we could gain about 33 % of memory.

The first problem can be corrected by not providing `dtype` or doing something similar to
```
dtype = np.int8 if _local_state.n_states < 2**7 else np.int32
self._local_states.states_to_numbers(x, dtype=dtype)
```
and the second with an integer division `//` since x should be an integer factor of `self.step`. Here, I am not sure how safe this is in the case when `self.start` and `self.step` are floats.

However, perhaps the current behaviour is desired?
A trivial change
```
StaticRange(-1, step=2, length=2, dtype=np.int8) -> StaticRange(-1/2, step=1, length=2)
```
would cause a memory issues as experienced. In the end, full sum calculations will eventually run out of memory for some `L` anyway.

¹ AFAIK, `to_sparse()` is not parallelized, so parallelization was useless.

**Comment by PhilipVinc** [comment]
Hey @macekmar , thanks for raising this point! Indeed, we might have overlooked this...

Ill ping @inailuig who is responsible for a lot of those changes to hear what he can contribute...

> In `HomogeneusHilbert.states_to_local_indices` ([L202](https://github.com/netket/netket/blob/781f2d2bd3b3eda6ebb6eb13469fa78aed9d20f5/netket/hilbert/homogeneous.py#L202))we do
> ```
> self._local_states.states_to_numbers(x, dtype=np.int32)
> ```
> which in case of `SpinHilbert` transforms an `int8` into `int32` array (x4 memory)

Yes, you are right. This is out of pure laziness. 
We should pick the smallest integer type that can fit the local hilbert dimension. This can be anything from `np.bool` to `np.int8`. In rare cases, `np.int16`.

> Next, in `StaticRange.states_to_numbers` ([L152](https://github.com/netket/netket/blob/781f2d2bd3b3eda6ebb6eb13469fa78aed9d20f5/netket/utils/static_range.py#L152)) we > do
> ```
> idx = (x - self.start) / self.step
> ```
> which transforms array `x` of type `int32` to `float64` (x8 memory than the original array). Note, that arrays `x` are different > arrays in different scopes.

Again, you are right. This code should/could be written to be more 'type stable' and parsimonious with promotion. 
The reason it is written like that was to support both jax and numpy array inputs, but it can be much improved...

I would support rewriting it in such a way to manually promote everything to the smallest possible type that can support both `x` and `self.start`.

> The first problem can be corrected by not providing dtype or doing something similar to

I'm not sure you can. Because if the local type is floating-point like (which is probably wrong, but happens due to historical reasons throughout netket) then the type remains floating point.
Maybe we can just enforce a more lax 'integer return type' in [here](https://github.com/netket/netket/blob/781f2d2bd3b3eda6ebb6eb13469fa78aed9d20f5/netket/utils/static_range.py#L153) of StaticRange, and that would solve the issue...

> and the second with an integer division // since x should be an integer factor of self.step. Here, I am not sure how safe this is in the case when self.start and self.step are floats.

How does // round floating point values?
Otherwise that's a good point.

As I said above, we should just stop supporting floating point samples... 
Mainly, if a change passes all the battery of tests in hilbert.py and `test/operators` it means it's probably ok...
Otherwise we can discuss.

I would greatly appreciate a PR with the fixes your proposed, and what I mention here.
It would allow us to see if the tests still pass.

**Comment by gcarleo** [comment]
we need floating point samples for continuous space, so if you want drop support for floating point in all samplers (personally I think it's maybe not a good idea) then one would need to treat the case of type support for continuous space samples separately....

**Comment by PhilipVinc** [comment]
Here we are discussing discrete spaces

**Comment by gcarleo** [comment]
> As I said above, we should just stop supporting floating point samples...

🤔

**Comment by gcarleo** [comment]
just out of curiosity, in what scenario is saving the full states the memory bottleneck rather than storing, e.g. the floating point amplitudes of a quantum states for ED?

**Comment by gcarleo** [comment]
yes I understand how... but the real saving is to chunk the calculation of the hamiltonian matrix elements rather than building all the hilbert space states at once? this should allow avoid saving the "samples"

**Comment by macekmar** [comment]
`StaticRange.length` determines the largest needed integer. So, the integer type could be determined when StaticRange is created. Then in `self._local_states.states_to_numbers(x, dtype=np.int32)` one does not have to provide a dtype.

Division // is a floor division (`__floordiv__`), so contrary to my first comment, it is not so trivial because the difference of _floats_ may be negative:
```
np.array(-ε//3).astype(np.int32) == -1
np.array(-ε/3).astype(np.int32) == 0
```
If we keep float `start` and `step` the same problem may arise: for some input $x$ we get but for $n+1$ we could get from
```
idx = (x - self.start) / self.step
```
$n + 1 - \epsilon$ which is then, after casting to integer, again $n$. This has a prospect being a well hidden error. 
IMHO, I may be panicking too much about floats: both start and step are always factors of 1/2 or integers, right? This should be fine even with floats (but raises the question if we really need floats at all).

BTW, in the docstring for StaticRange: `... configurations to integer indices ∈ [0,length].` it should be `[0,length)`. Also, `StaticRange.numbers_to_states` works for indices outside `[0,length)`. I guess it should raise an error, right?

@gcarleo : About your comment if memory issues are relevant. 
Samples $\sigma$ and $\Psi(\sigma)$ store $N \cdot 2^N$ and $2^N$ numbers. Similarly, the sparse matrix stores $2^N \cdot n_{\mathrm{conn}}$ of indices and values. I think Lanczos only stores additional $\Psi(\sigma)$-sized vectors.
But here the problem arises that we are manipulating $N \cdot 2^N \cdot n_{\mathrm{conn}}$.

**Comment by PhilipVinc** [comment]
> IMHO, I may be panicking too much about floats: both start and step are always factors of 1/2 or integers, right? This should be fine even with floats (but raises the question if we really need floats at all).

exactly my point. I think the fact that we consistently work with 1/2 or integers steps is ‘saving’ us but indeed we could just use integers and be safe.

To do that, however, we’d need to merge some form of #1938.

In any case for the time being I’m ok to ‘ignore’ potential issues if tests pass. They are very extensive so I trust them.


So, a PR might just auto set a default dtype in the StaticRange and this should be enough?
Feel free to address those docstring issues as well

**Comment by PhilipVinc** [comment]
This should now be addressed in master
